src/plugins/data/server/search/routes/msearch.ts

```diff
@@ -0,0 +1,136 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { first } from 'rxjs/operators';
+import { schema } from '@kbn/config-schema';
+
+import { IRouter } from 'src/core/server';
+import { UI_SETTINGS } from '../../../common';
+import { SearchRouteDependencies } from '../search_service';
+import { getDefaultSearchParams } from '..';
+
+interface MsearchHeaders {
+  index: string;
+  preference?: number | string;
+}
+
+interface MsearchRequest {
+  header: MsearchHeaders;
+  body: any;
+}
+
+interface RequestBody {
+  searches: MsearchRequest[];
+}
+
+/** @internal */
+export function convertRequestBody(
+  requestBody: RequestBody,
+  { timeout }: { timeout?: string }
+): string {
+  return requestBody.searches.reduce((req, curr) => {
+    const header = JSON.stringify({
+      ignore_unavailable: true,
+      ...curr.header,
+    });
+    const body = JSON.stringify({
+      timeout,
+      ...curr.body,
+    });
+    return `${req}${header}\n${body}\n`;
+  }, '');
+}
+
+/**
+ * The msearch route takes in an array of searches, each consisting of header
+ * and body json, and reformts them into a single request for the _msearch API.
+ *
+ * The reason for taking requests in a different format is so that we can
+ * inject values into each request without needing to manually parse each one.
+ *
+ * This route is internal and _should not be used_ in any new areas of code.
+ * It only exists as a means of removing remaining dependencies on the
+ * legacy ES client.
+ *
+ * @deprecated
+ */
+export function registerMsearchRoute(router: IRouter, deps: SearchRouteDependencies): void {
+  router.post(
+    {
+      path: '/internal/_msearch',
+      validate: {
+        body: schema.object({
+          searches: schema.arrayOf(
+            schema.object({
+              header: schema.object(
+                {
+                  index: schema.string(),
+                  preference: schema.maybe(schema.oneOf([schema.number(), schema.string()])),
+                },
+                { unknowns: 'allow' }
+              ),
+              body: schema.object({}, { unknowns: 'allow' }),
+            })
+          ),
+        }),
+      },
+    },
+    async (context, request, res) => {
+      const client = context.core.elasticsearch.client.asCurrentUser;
+
+      // get shardTimeout
+      const config = await deps.globalConfig$.pipe(first()).toPromise();
+      const { timeout } = getDefaultSearchParams(config);
+
+      const body = convertRequestBody(request.body, { timeout });
+
+      try {
+        const ignoreThrottled = !(await context.core.uiSettings.client.get(
+          UI_SETTINGS.SEARCH_INCLUDE_FROZEN
+        ));
+        const maxConcurrentShardRequests = await context.core.uiSettings.client.get(
+          UI_SETTINGS.COURIER_MAX_CONCURRENT_SHARD_REQUESTS
+        );
+        const response = await client.transport.request({
```

Comment:
Do you think we could get away with using `client.msearch` rather than `transport.request` here? In other places the lack of typings has limited us but it might be worth a shot.

Reply:
Also, can we add some logic for aborting requests that have disconnected? Here are the relevant bits of code:

https://github.com/elastic/kibana/blob/a3af33ddaf32cfbfb53f66478bf20649a998a06b/src/plugins/data/server/search/routes.ts#L45

https://github.com/elastic/kibana/blob/4a69b5145ee5def5b2480bb4f592e6d5987fbba1/src/plugins/data/server/search/es_search/es_search_strategy.ts#L50-L52

Reply:
I've added quite a few changes to this code in my PR.
Better I add it in mine, to avoid conflicts

Reply:
Okay then if that works for you @lukasolson, I'll plan to merge this as soon as I get the index patterns / rollups thing sorted out.